<format>: Add format, vformat, and local-less vformat_to

eldakesh-ms opened the pull request:
Adds the functions. Only locally tested, as I do not want to mess up the
merge of the testing overhaul PR (#1719). These will be tested through
`format`, since it is just a pass-through.

`vformat_to` uses `locale::classic()` always, which is probably the right behavior
but we may want to change `_Format_handler` to hold a pointer instead of
a reference to a locale.

<!--
Before submitting a pull request, please ensure that:

* Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.

* These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).

* Your changes are written from scratch using only this repository, the C++
  Working Draft (including any cited standards), other WG21 papers (excluding
  reference implementations outside of proposed standard wording), and LWG
  issues as reference material. If your changes are derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If your changes are derived from any other project, you *must* mention it
  here, so we can determine whether the license is compatible and what else
  needs to be done.
-->

CaseyCarter commented on stl/inc/format:
```diff
@@ -1376,11 +1392,49 @@ _OutputIt vformat_to(
-wstring vformat(const locale& _Loc, wstring_view _Fmt, wformat_args _Args);
+string vformat(string_view _Fmt, format_args _Args) {
+    string _Str;
+    vformat_to(back_inserter(_Str), _Fmt, _Args);
```
`vformat_to` and `back_inserter` both need `_STD` qualification. (Several places.)

eldakesh-ms replied:
I was under the understanding that there can't be ADL hijacking since these are types within the standard. Is there a scenario when this is unsafe?

CaseyCarter replied:
We conventionally `_STD`-qualify pretty names regardless so it's obvious even to someone passing by and glancing at this code that we didn't forget to qualify. But yes, I agree these arguments are of types that we fully control so ADL hijacking isn't possible here.

eldakesh-ms replied:
I can still add the qualifier, I just wanted to make sure my understanding of when `_STD` is required as opposed to when it is stylistic.

CaseyCarter commented on stl/inc/format:
```diff
@@ -1376,11 +1392,49 @@ _OutputIt vformat_to(
+
+template <class... _Types>
+string format(string_view _Fmt, const _Types&... _Args) {
+    return vformat(_Fmt, _STD make_format_args(_Args...));
```
Ditto `vformat` needs `_STD` qualification. (Several places.)

eldakesh-ms commented:
The tests are failing but I am unable to repro them. Any ideas?

CaseyCarter commented on stl/inc/format:
```diff
@@ -1376,11 +1392,49 @@ _OutputIt vformat_to(
-wstring vformat(wstring_view _Fmt, wformat_args _Args);
-string vformat(const locale& _Loc, string_view _Fmt, format_args _Args);
-wstring vformat(const locale& _Loc, wstring_view _Fmt, wformat_args _Args);
+string vformat(string_view _Fmt, format_args _Args) {
```
Missing `inline` (Four places.)

eldakesh-ms commented:
Ok so there is a fix in #1719 for the `_Write(bool)` function to use a wide string when necessary. I will wait until that goes in (and add tests to these functions).

eldakesh-ms commented:
Rebased onto `feature/format` because #1719 got merged (thanks @barcharcraz!) I also added a macro to make the string testing a bit less gross. I didn't convert the whole file because I want to get feedback on this approach first, however I think it improves readability a lot, and since it's a test I am not too upset about a macro like that.

CaseyCarter reviewed (commented):
This looks fine to me, but @barcharcraz's eyes turn red when he sees preprocessor use so :shrug:

barcharcraz commented:
> Rebased onto `feature/format` because #1719 got merged (thanks @barcharcraz!) I also added a macro to make the string testing a bit less gross. I didn't convert the whole file because I want to get feedback on this approach first, however I think it improves readability a lot, and since it's a test I am not too upset about a macro like that.

I like it, even though it makes the macro much, much less hygenic (you're relying on the type parameter name being charT)
